File: src/walkway_perception/scripts/clear_costmap_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav2_msgs.srv import ClearEntireCostmap

class ClearCostmapClient(Node):
    def __init__(self):
        super().__init__('clear_costmap_client')
        # Initialize client without blocking the constructor
        self.local_client = self.create_client(
            ClearEntireCostmap, 
            '/local_costmap/clear_entirely_local_costmap'
        )

        self.global_client = self.create_client(
            ClearEntireCostmap,
            '/global_costmap/clear_entirely_global_costmap'
        )

    def send_clear_request(self):
        # Implement a safe timeout so the node doesn't hang indefinitely
        if not self.local_client.wait_for_service(timeout_sec=3.0):
            self.get_logger().error('Costmap service not available after 3 seconds.')
            return False

        request = ClearEntireCostmap.Request()
        self.get_logger().info('Sending request to clear costmaps...')
        
        # Store the future as an instance attribute so we can spin on it in main
        self.local_future = self.local_client.call_async(request)
        self.global_future = self.global_client.call_async(request)
        return True

def main(args=None):
    rclpy.init(args=args)
    node = ClearCostmapClient()

    try:
        if node.send_clear_request():
            # Spin the executor safely until the future is resolved
            rclpy.spin_until_future_complete(node, node.local_future)
            
            if node.local_future.result() is not None:
                node.get_logger().info('Local costmap cleared successfully!')
            if node.global_future.result() is not None:
                node.get_logger().info('Global costmap cleared successfully!')
            else:
                node.get_logger().error('Service call completed, but returned an empty/failed response.')
    
    except KeyboardInterrupt:
        node.get_logger().info('Manual interruption detected. Shutting down.')
    except Exception as e:
        node.get_logger().error(f'An unexpected error occurred: {e}')
        
    finally:
        # Guarantee clean teardown of the ROS 2 context
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/walkway_perception/src/clear_costmap_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "nav2_msgs/srv/clear_entire_costmap.hpp"

#include <chrono>
#include <memory>

using namespace std::chrono_literals;

class ClearCostmapClient : public rclcpp::Node {
    public:
    ClearCostmapClient() : Node("clear_costmap_client") {
        client_ = this->create_client<nav2_msgs::srv::ClearEntireCostmap>(
            "/local_costmap/clear_entirely_local_costmap"
        );
    }

    // Returns a shared future so the main thread can spin on it
    std::shared_future<std::shared_ptr<nav2_msgs::srv::ClearEntireCostmap::Response>> send_request() 
    {
        if (!client_->wait_for_service(3s)) {
            RCLCPP_ERROR(this->get_logger(), "Costmap service not available.");
            // Return an invalid future to indicate failure
            return {};
        }

        auto request = std::make_shared<nav2_msgs::srv::ClearEntireCostmap::Request>();
        RCLCPP_INFO(this->get_logger(), "Sending request to clear the local costmap...");

        // Asynchronously send the request
        return client_->async_send_request(request).future.share();
    }

    private:
    rclcpp::Client<nav2_msgs::srv::ClearEntireCostmap>::SharedPtr client_;
};

int main(int argc, char **argv) 
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ClearCostmapClient>();

    auto future = node->send_request();

    // Check if the future is valid
    if (future.valid()) {
        // Spin the node until the future completes
        if (rclcpp::spin_until_future_complete(node, future) == rclcpp::FutureReturnCode::SUCCESS) {
            RCLCPP_INFO(node->get_logger(), "Costmap cleared successfully!");
        } else {
            RCLCPP_ERROR(node->get_logger(), "Failed to call clear costmap service.");
        }
    } else {
        RCLCPP_ERROR(node->get_logger(), "Service call was not initiated. Shutting down.");
    }

    rclcpp::shutdown();
    return 0;
}